quinn-proto/src/connection.rs

```diff
@@ -1086,26 +1081,28 @@ impl Connection {
                 }
                 let closed =
                     self.process_payload(mux, now, number.unwrap(), packet.payload.into())?;
-                Ok(if closed {
+                self.state = if closed {
                     State::Draining
                 } else {
                     State::Established
-                })
+                };
+                Ok(())
             }
-            State::Closed(state) => {
+            State::Closed(_) => {
                 for frame in frame::Iter::new(packet.payload.into()) {
                     match frame {
                         Frame::ConnectionClose(_) | Frame::ApplicationClose(_) => {
                             trace!(self.log, "draining");
-                            return Ok(State::Draining);
+                            self.state = State::Draining;
+                            return Ok(());
                         }
                         _ => {}
                     }
                 }
-                Ok(State::Closed(state))
+                Ok(())
             }
-            State::Draining => Ok(State::Draining),
-            State::Drained => Ok(State::Drained),
+            State::Draining => Ok(()),
+            State::Drained => Ok(()),
```

Comment:
These could be folded together into `State::Draining | State::Drained => Ok(())`

Reply:
Nice.